Add support for execution via host file descriptor

blechschmidt opened the pull request:
Hi,

this commit adds support for program execution via a host file descriptor. To make use of this feature, the host file descriptor must be mapped to -1 in the guest. For example,

    exec 3</usr/bin/echo
    runsc exec --pass-fd=3:-1 mycontainer hello world

will run the host's echo binary inside gVisor. In this case, "hello" is supplied to echo as argv[0]. As a result, the output of the above command is "world".

This feature is useful for bootstrapping unknown guest environments and allows static executables to perform setup actions inside the container while they need not be part of the guest file system.

Again, I am not sure whether this is a feature you would agree on merging. If not, please feel free to close this PR. Otherwise, I am happy about feedback on the implementation details (for example in case you prefer an extra command line argument instead of mapping to the special FD value of -1 through the CLI).

nlacasse commented:
This is very cool!  We have something similar internally, which is why `kernel.ProcArgs.File` exists (which you are using correctly).  So I totally understand the utility of this feature.

I think I would prefer a distinct command line flag for this.  It would make it easier to document the functionality.  And something feels wrong about using "-1" as an FD.

How do you feel about an `--exec-fd=N` flag for `run` and `exec`?

I left some other small comments, but overall the approach looks great!  Thank you.

blechschmidt commented:
Thank you for the feedback.

> How do you feel about an `--exec-fd=N` flag for `run` and `exec`?

This is what I had in mind as an alternative to the current approach.

> I left some other small comments

Could you please look into this again? My GitHub interface currently does not display your comments or a code review.

nlacasse reviewed (changes requested):
Sorry I forgot to hit the review button.

blechschmidt commented on pkg/sentry/control/proc.go:
```diff
@@ -234,6 +229,28 @@ func (proc *Proc) execAsync(args *ExecArgs) (*kernel.ThreadGroup, kernel.ThreadI
+		defer file.DecRef(ctx)
+		execFD.Release()
+		initArgs.File = file
+		initArgs.Filename = ""
```
An error is returned in such a case now.

https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/control/proc.go#L244-L247

blechschmidt commented on runsc/boot/vfs.go:
```diff
@@ -172,13 +172,16 @@ func registerFilesystems(k *kernel.Kernel) error {
 
+	if !resolveExecutable {
+		return nil
+	}
```
You are right. I have replaced this with the following:

https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/runsc/boot/vfs.go#L182-L185

blechschmidt commented on runsc/boot/loader.go:
```diff
@@ -852,14 +852,30 @@ func (l *Loader) startSubcontainer(spec *specs.Spec, conf *config.Config, cid st
+		execFD.Release()
+
+		info.procArgs.File = file
+		info.procArgs.Filename = ""
```
This errors now as well:

https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/runsc/boot/loader.go#L873-L876

blechschmidt commented:
Thank you again for the review. I have implemented the requested changes and changed all locations in which an FD of -1 was mapped. `exec-fd` is now decoupled from `pass-fd`.

avagin commented on runsc/container/container_test.go:
```diff
@@ -2892,3 +2899,136 @@ func TestFDPassingExec(t *testing.T) {
+	}
+
+	// Open the echo binary as a file.
+	echoFile, err := os.Open(echoPath)
```
echo can depend on libraries and runsc will look them up on the container file system:
$ ldd /bin/echo
	linux-vdso.so.1 (0x00007ffc806f3000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8cb7aa0000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f8cb7caa000)

I think you need to add a static linked test binary for this test.

blechschmidt replied:
Is that a problem? The file system is shared in the test.

avagin replied:
nm. I was thinking we use a custom image here.

avagin commented:
Is it possible to open /proc/pid/exe of a process executed via a host fd? Is it possible to rewrite a binary via /proc/pid/exec?

blechschmidt commented:
Reading `/proc/pid/exe` works. As you would not pass a file descriptor opened for writing, I don't see how that could happen. Operations on the file descriptor are still guarded by host permissions.

blechschmidt commented:
To be on the safe side, I have looked into the security implications of this feature again in detail.

1. As the file descriptor is not imported into the file descriptor table of any process, it cannot be accessed through `/proc/<pid>/fd/<fd>`. This leaves us with `/proc/<pid>/exe`.

2. In general, I assume that an FD for execution is opened as `O_RDONLY`.

3. Access to `/proc/self/exe` is guarded by ptrace access restrictions:
https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/fsimpl/proc/task_files.go#L1003-L1007

For `/proc/self/exe`, access is prevented through `canTraceYAMALocked`:
https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/kernel/ptrace.go#L126-L132

The reason is that none of the conditions, which are also outlined in the comments of the function, match:
https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/kernel/ptrace.go#L250-L266
By default, these conditions are never true for another process in relation to a process started through `exec-fd`.

Yama is currently hardcoded to always be enabled for ptrace access checking:
https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/kernel/kernel.go#L415

Thus, with Yama enabled, there is currently no way to modifiy the host executable.

4. Assuming that Yama was disabled, operations on the host file descriptor would be mediated by the implementation in `pkg/sentry/fsimpl/host/host.go`. In such a case, we are still left to check for some dangerous operations. Thanks to the fact that these operations are mediated, we only need to check operations that are implemented there and potentially modify the state of the host file.

    1. The first function that requires investigation is `SetStat`. `fchmod`, which counterintuitively also works with `O_RDONLY` descriptors, is disallowed. So this is not an issue.
       https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/fsimpl/host/host.go#L532-L540
       
       Setting the file size performs an `Ftruncate` call on the host. Since `Ftruncate` only works with file descriptors opened for writing, this is okay, too.
       https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/fsimpl/host/host.go#L541-L556
       
       Setting access and modification times results in a `utimensat` call on the host. To put it short, these two times could be changed on the host if Yama was disabled, in particular when running a rootless containers and with an `O_RDONLY` file descriptor.
       https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/fsimpl/host/host.go#L557-L565
    
    2. `Fallocate` is simply called on the host. It requires that the file descriptor is open for writing. So this is fine, too.
    https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/fsimpl/host/host.go#L714-L716
    
    3. We have the `Pwrite` and `Write` functions that use the `writeToHostFD` functions to write and sync changes to the host. As write operations require that the FD is opened for writing, this is okay as well.
    
    4. `Fsync` does not require `O_RDONLY`, but it is also ineffective if no preceding host-changing operations take place.
    https://github.com/google/gvisor/blob/998055564792588d8396ee7eb5ba0682e118549e/pkg/sentry/fsimpl/host/host.go#L919-L923

Based on these insights, I have implemented the following modifications, so that this feature does not rely on Yama or ptrace access restrictions in general:

1. Through an additional security check, it is ensured that only `O_RDONLY` file descriptors end up being used for execution.
https://github.com/google/gvisor/blob/761bda09a55eed621673e9742fe5c5747fa42677/pkg/sentry/fsimpl/host/host.go#L240-L255

2. `Readonly` and `readonly` members are introduced to the `NewFDOptions` and `inode` structs in `pkg/sentry/fsimpl/host/host.go`. In case `inode.readonly` is false, all operations that may potentially change the file on the host are refused with `EPERM` (i.e. the operations listed above). Strictly, this is not necessary in all cases, but it has the advantages that access checks already take place in runsc instead of the host and it reduces the number of syscalls to the host.

nlacasse commented:
This is merged.  Thank you!

blechschmidt commented:
Awesome! Thanks a lot.
